### PR TITLE
Handle disabled display within ensure_unlocked_desktop

### PR DESCRIPTION
**Pull request by foursixnine**
As a consequence of bsc#1168979 and possibly boo#1168992 when the SUT's
power saving kicks in, the screen buffer is not properly updated once it
comes back, causing openQA to see the wrong screen, upon a keypress the
SUT ends up believing that it should expect i.e generic desktop, when in
reality the screen is turned off (because again, the power saving mode
is just kicking in), this causes openQA to detect the wrong screen and
fail the test, when actually the system is still working.

Matching the guest disabled display screen, helps us to detect these
situations, avoiding entering into the wrong loop (see
https://openqa.opensuse.org/tests/1259244#step/consoletest_finish/13)

Sometimes the screenbuffer does not refresh, to workaround this
issue we now send 'Shift' to the SUT

Test is simply logging into user console, waiting for either 600 or 300 
seconds and then switching back, only to find (sometimes, when the time is 
600 secs) a screen with old buffer data (that matches generic desktop)

See: https://progress.opensuse.org/issues/64812

Verifications: https://openqa.suse.de/tests/overview?groupid=96&build=foursixnine_189.1&distri=sle&version=15-SP2&flavor=black-display-workaround-now-finally-really-finally

### CODE REVIEW

File: lib/x11utils.pm [hunk 1]
```diff
@@ -86,6 +88,12 @@ sub ensure_unlocked_desktop {
                 select_user_gnome($username);
             }
         }
+        if (match_has_tag('guest-disabled-display')) {
```
Comment (foursixnine) [inline lib/x11utils.pm:91]: @ggardet this correspond to the needle that you created recently, but as screenlock, I'll submit a pr later to do the respective change there. I hope the intention here is clear. 

I'll submit later also a VR for opensuse (In hopes that it behaves the same way too)

File: lib/x11utils.pm [hunk 2]
```diff
@@ -112,6 +120,8 @@ sub ensure_unlocked_desktop {
                     assert_screen 'generic-desktop';
                 }
                 else {
+                    diag("Next loop ($counter), Generic desktop didn't match");
+                    record_soft_failure('bsc#1168979') if is_aarch64;
```
Comment (foursixnine) [inline lib/x11utils.pm:124]: @SergioAtSUSE this is the closes to the workaround that you wanted, that I find myself to agree to.

File: lib/x11utils.pm [hunk 3]
```diff
@@ -1,4 +1,4 @@
-# Copyright (C) 2019 SUSE LLC
+# Copyright (C) 2020 SUSE LLC
```
Comment (SergioAtSUSE) [inline lib/x11utils.pm:1]:
```suggestion
# Copyright (C) 2019-2020 SUSE LLC
```

File: lib/x11utils.pm [hunk 4]
```diff
@@ -86,6 +88,12 @@ sub ensure_unlocked_desktop {
                 select_user_gnome($username);
             }
         }
+        if (match_has_tag('guest-disabled-display')) {
+            wait_screen_change(sub {
+                    send_key 'shift';
+            }, 10);
+            record_info('Guest disabled display', 'Might be consequence of bsc#1168979');
```
Comment (SergioAtSUSE) [inline lib/x11utils.pm:95]: Since bugzilla is not available, I cannot see the content of bsc#1168992.
I suppose that you are addressing bsc#1168992 and that is different from bsc#1168979

This is wrong, it is not detecting bsc#1168979 unequivocally.


```suggestion
            record_info('Guest disabled display', 'Might be consequence of bsc#1168992');
```

File: lib/x11utils.pm (hunk 2 above)
Comment (SergioAtSUSE) [reply lib/x11utils.pm:124]: This is wrong, the `esc` key is sent before, causing the display to redraw. bsc#1168979 causes this branch to trigger, but it is a fake. When bsc#1168979 is reproduced, the screen is locked, we should handle this before the the `esc` key. And `esc` triggers a different behavior from `shift`.

`shift` just causes the screen to redraw.

`esc` causes the lock image to roll up and show the unlock dialog.

The problem with `esc` is that the animation takes some time, in this overloaded environment and aarch64 it is too much time wasted for a workaround to detect a bug. `shift` should be used.

You can remove this part so it can be handled better in #10064

**Comment by foursixnine** [comment]
> > As a consequence of bsc#1168979 and possibly boo#1168992 when the SUT's
> > power saving kicks in, the screen buffer is not properly updated once it
> > comes back, causing openQA to see the wrong screen, upon a keypress the
> > SUT ends up believing that it should expect i.e generic desktop, when in
> > reality the screen is turned off (because again, the power saving mode
> > is just kicking in), this causes openQA to detect the wrong screen and
> > fail the test, when actually the system is still working.
> 
> I cannot see bugzilla, so I cannot check boo#1168992, but for bsc#1168979, that is wrong.
> 
> The power saving does not cause the outdated screen buffer of bsc#1168979.
> The outdated screen buffer makes isotovideo to fail, waiting for needle desktop-runner until it times out. And while isotovideo is waiting for the needle desktop-runner, the power saving kicks in, showing a "black screen" later.
> 
> The power saving happens after the bug, not before. The black screen is a consequence of the timeout of isotovideo caused by the bug.
> The bug is not caused by power saving and it doesn't happen before.

As discussed: this is aftermath of output being outdated. 

> > Matching the guest disabled display screen, helps us to detect these
> > situations, avoiding entering into the wrong loop (see
> > https://openqa.opensuse.org/tests/1259244#step/consoletest_finish/13)
> 
> The handling is good for the situations when the display is off, but it is not a valid workaround for bsc#1168979 and it is not detecting bsc#1168979 unequivocally.

I'm not aiming to detect the outdated screen buffer (Because ensure_unlock_desktop has no idea what the last screen visible in X11 console was) I just want to ensure, that the test doesn't die, as a consequence of: product bug, or power saving having kicked in.

> > Test is simply logging into user console, waiting for either 600 or 300
> > seconds and then switching back, only to find (sometimes, when the time is
> > 600 secs) a screen with old buffer data (that matches generic desktop)
> 
> Nice! I think we should add this test module to the system_performance test suite. Or it could even have its own test suite.

No, the test module is just there as a test, and to help verifying further changes, investing time in adding this as a test module, is a waste of resources (All this together, is already a waste of the time for many of us) 

> You can remove this part so it can be handled better in #10064

See https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/10064/#discussion_r421806856 https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/10064/#discussion_r419393118

I guess what we can do, is change the `esc` for a `shift` in L109.

**Comment by SergioAtSUSE** [comment]
> I'm not aiming to detect the outdated screen buffer (Because ensure_unlock_desktop has no idea what the last screen visible in X11 console was) I just want to ensure, that the test doesn't die, as a consequence of: product bug, or power saving having kicked in.

1. We have already discussed a lot of times that **workaround** means avoiding a test to fail because a product bug, but it is wrong to workaround as green.
2. Each workaround must cause a soft-fail and each soft-fail must be recorded by a workaround
3. A soft-fail should only be recorded if we have properly detected the bug.

So, please, don't hide bsc#1168979 and don't record a soft fail in the wrong place for bsc#1168979.


> No, the test module is just there as a test, and to help verifying further changes, investing time in adding this as a test module, is a waste of resources (All this together, is already a waste of the time for many of us)

I disagree. I don't consider reproducing a bug and giving more data to developers to fix it a waste of time.
Discussions on disagreement is not a waste of time, different point of views are expressed. That a task takes more time than plan is not a waste of time is bad planning.


> I guess what we can do, is change the `esc` for a `shift` in L109.

Since we still don't agree on how to handle bsc#1168979, please leave the handling of bsc#1168979 for later in #10064 so we can at least handle the problem you are trying to address.

**Comment by foursixnine** [comment]
> I'm not aiming to detect the outdated screen buffer (Because ensure_unlock_desktop has no idea what the last screen visible in X11 console was) I just want to ensure, that the test doesn't die, as a consequence of: product bug, or power saving having kicked in.
> 
As for detecting the old outdated screen buffer: 

1. See the clock: https://openqa.suse.de/tests/4211435#step/boot_to_desktop/9
2: See the clock: https://openqa.suse.de/tests/4211435#step/desktop_stuff/12, `generic-desktop` matches, from the **OLD** (See line above) screen, test presses `esc` and then `alt-f2` to get the desktop runner, since next `check_screen` is for `desktop-runner`, fails and the `else` condition in the code kicks in
3: Since there's nothing to do, the else jumps next, in my code changes, a `record_soft_failure` is triggered  https://openqa.suse.de/tests/4211435#step/desktop_stuff/14 in this case, specifically for aarch64 (but tbh, it is not needed, since it could happen somewhere else). In the next run of the loop, power saving kicks in (because 600 seconds had passed) and screen is turned black[1], at screenshot 15, guest display is detected as turned off ([1]) and the shift is sent, ending up in the password prompt, which ultimately succeeds
4: See the clock: https://openqa.suse.de/tests/4211435#step/desktop_stuff/19 

And here you go: Bug + handling of it's consequences.

You can refer to the autoinst-log.txt and follow the traces to double check what's going on. 

This is independant of the worker load (only that it might take more time to execute scripts in other consoles, causing the screen lock in the first place).

In any case, with regards to [1], I think that it might also be that a screen that is locked, will turn off after some seconds anyway, which is why (or my theory) on why the guest screen is disabled.

File: lib/x11utils.pm [hunk 5]
```diff
@@ -68,8 +69,9 @@ all possible options should be handled within loop to get unlocked desktop
 =cut
 sub ensure_unlocked_desktop {
     my $counter = 10;
+
     while ($counter--) {
-        my @tags = qw(displaymanager displaymanager-password-prompt generic-desktop screenlock screenlock-password authentication-required-user-settings authentication-required-modify-system);
+        my @tags = qw(displaymanager displaymanager-password-prompt generic-desktop screenlock screenlock-password authentication-required-user-settings authentication-required-modify-system guest-disabled-display);
         push(@tags, 'blackscreen') if get_var("DESKTOP") =~ /minimalx|xfce/;    # Only xscreensaver and xfce have a blackscreen as screenlock
```
Comment (okurz) [inline lib/x11utils.pm:75]: IIUC "blackscreen" and "guest-disabled-display" serve the same purpose. So far this really seems to be aarch64 specific so I suggest to only enable there, is it correct? Maybe it's because of running a more recent qemu on our aarch64 then of course enable it for all.

File: lib/x11utils.pm (hunk 4 above)
Comment (okurz) [reply lib/x11utils.pm:95]: @SergioAtSUSE the record info is a weak reference to the bug, not stating that it would *detect* it. I am ok with this code as is.

File: lib/x11utils.pm (hunk 2 above)
Comment (okurz) [reply lib/x11utils.pm:124]: I also think that this is most likely not working properly as the "esc" key was pressed just in before so only the *next* loop cycle should try to detect a bug condition

File: lib/x11utils.pm [hunk 6]
```diff
@@ -121,6 +131,7 @@ sub ensure_unlocked_desktop {
         if (match_has_tag('screenlock') || match_has_tag('blackscreen')) {
             wait_screen_change {
                 send_key 'esc';                                                                           # end screenlock
+                record_info("Screen lock present");
```
Comment (okurz) [inline lib/x11utils.pm:134]: IMHO an overuse of `record_info`. Common operations like matching on a screen or pressing a key should not be accompanied with a `record_info`.

File: tests/foursixnine/desktop_stuff.pm [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+use strict;
+use warnings;
+use base "opensusebasetest";
+use testapi;
+use utils;
+use x11utils 'ensure_unlocked_desktop';
+
+sub run {
+    select_console('user-console');
+    script_run('for key in $(gsettings list-keys org.gnome.desktop.session); do echo $key $(gsettings get org.gnome.desktop.session $key); done');
```
Comment (okurz) [inline tests/foursixnine/desktop_stuff.pm:30]: As you are already reading out the settings you could set a very short idle time and only sleep for that time. If you want to be strict you can assert that the previously set value is a sane default

File: lib/x11utils.pm (hunk 5 above)
Comment (ggardet) [reply lib/x11utils.pm:75]: I think this is more virtio specific.

File: lib/x11utils.pm (hunk 1 above)
Comment (ggardet) [reply lib/x11utils.pm:91]: I doubt that hitting the `shift` key brings the screen back. Does it?

File: lib/x11utils.pm (hunk 5 above)
Comment (okurz) [reply lib/x11utils.pm:75]: even better to use that as the condition

File: tests/foursixnine/desktop_stuff.pm (hunk 7 above)
Comment (foursixnine) [reply tests/foursixnine/desktop_stuff.pm:30]: Nah, this test module, is actually for testing purposes

File: lib/x11utils.pm (hunk 1 above)
Comment (foursixnine) [reply lib/x11utils.pm:91]: According to @SergioAtSUSE it does (And in general, looking at the test runs, it seems to do, I'll trigger a test later and see manually)

File: lib/x11utils.pm (hunk 2 above)
Comment (foursixnine) [reply lib/x11utils.pm:124]: See https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/10217#issuecomment-625882432 the `esc` does work, but it only causes the screen to be redrawn, hence the screen lock being detected in the next step (where `desktop-runner` is actually expected). One thing that comes to my mind, is pressing shift right at the beginning of the loop (or even out of it), but in doing so... we _miss/hide_ the bug, so I'd be against that

File: lib/x11utils.pm (hunk 1 above)
Comment (ggardet) [reply lib/x11utils.pm:91]: So, this is definitely not https://bugzilla.opensuse.org/show_bug.cgi?id=1168992

File: lib/x11utils.pm (hunk 4 above)
Comment (ggardet) [reply lib/x11utils.pm:95]: This is not https://bugzilla.opensuse.org/show_bug.cgi?id=1168992 as pressing `esc` or `shift` let us quit the "Guest disabled display" state.

Comment (foursixnine) [reply lib/x11utils.pm:95]: @ggardet thanks for confirming

File: lib/x11utils.pm (hunk 5 above)
Comment (foursixnine) [reply lib/x11utils.pm:75]: @okurz I've seen the guest-disabled-display somewhere else in x86. I'd say is more of a generic thing, and only in the aarch specific case, I'd tag it as a possible response to the bug in question, as @ggardet is already confirming that this is not the bug he's seeing on opensuse

File: lib/x11utils.pm (hunk 2 above)
Comment (SergioAtSUSE) [reply lib/x11utils.pm:124]: > the `esc` does work, but it only causes the screen to be redrawn

That is wrong.

- Any key will turn on the display.
- `shift`, `ctrl`, `alt` (maybe others, I haven't checked the whole keyboard): cause the screen to redraw, showing the full screen image
- `esc` or alphanumeric key: cause the screen to redraw **and to roll up the full screen image**, showing the unlock dialog.

@ggardet, please, test is with GNOME and GDM, this is the behavior since a long time.


> hence the screen lock being detected in the next step (where `desktop-runner` is actually expected).

To avoid misleading, let me rephrase it, because the test doesn't detect the screen lock (that is why it fails), a reviewer detects it.

"hence the screen lock being shown"


>  One thing that comes to my mind, is pressing shift right at the beginning of the loop (or even out of it), but in doing so... we _miss/hide_ the bug, so I'd be against that

I agree, any key sent at the beginning of the loop hides the bug.
But, the proposal in #10064 was to check the display before pressing the key. And the content of #10064 is a the result of changing my initial proposal of including the handling in the `generic-desktop` branch.

If the test is already inside the `generic-desktop` branch, then we can send `shift`, because we already detected a `generic-desktop`. After sending that key, if we detect the screen lock ("full image" or "unlock dialog"), the bug has being detected and a soft-fail can be recorded.
And, my

File: lib/x11utils.pm (hunk 4 above)
Comment (SergioAtSUSE) [reply lib/x11utils.pm:95]: > @SergioAtSUSE the record info is a weak reference to the bug, not stating that it would _detect_ it. I am ok with this code as is.

This will record a soft-fail anytime ensure_unlocked_desktop finds a turned off display, which is not always a consequence of bsc#1168979, How can you ok with it when it can be done right?

**Comment by SergioAtSUSE** [comment]
> As for detecting the old outdated screen buffer:
> 
>     1. See the clock: https://openqa.suse.de/tests/4211435#step/boot_to_desktop/9
>        2: See the clock: https://openqa.suse.de/tests/4211435#step/desktop_stuff/12, `generic-desktop` matches, from the **OLD** (See line above) screen, test presses `esc` and then `alt-f2` to get the desktop runner, since next `check_screen` is for `desktop-runner`, fails and the `else` condition in the code kicks in

I agree. I am telling this same thing since I created the ticket.

> 3: Since there's nothing to do, the else jumps next, in my code changes, a `record_soft_failure` is triggered  https://openqa.suse.de/tests/4211435#step/desktop_stuff/14 in this case, specifically for aarch64 (but tbh, it is not needed, since it could happen somewhere else).

the turned off display can happened somewhere else and for any other reason which is not bsc#1168979. So, please don't record a soft-fail for bsc#1168979 when bsc#1168979 is not reproduced.

Pick up the microphone again ;)

**Comment by SergioAtSUSE** [comment]
Has really nobody else commenting on this PR tested the behavior of GNOME and GDM?

**Comment by SergioAtSUSE** [comment]
@foursixnine, as you are concerned about the time invested (wasted?) in this ticket. Why don't better start working on https://progress.opensuse.org/issues/40490

When that ticket is done, wouldn't you consider the effort put in this PR a waste of time?

File: tests/foursixnine/desktop_stuff.pm (hunk 7 above)
Comment (okurz) [reply tests/foursixnine/desktop_stuff.pm]: Yeah I got that but we want to have an "explicit test for screensaver" nevertheless in the future, right? And I think a valid test would be:

* assert that screensaver is enabled by default with sane timeout
* set screensaver to low value N seconds
* assert screensaver activates after N seconds
* disable screensaver for good to let other test modules not care about screensaver at all

File: lib/x11utils.pm (hunk 4 above)
Comment (SergioAtSUSE) [reply lib/x11utils.pm:95]: > > @SergioAtSUSE the record info is a weak reference to the bug, not stating that it would _detect_ it. I am ok with this code as is.
> 
> This will record a soft-fail anytime ensure_unlocked_desktop finds a turned off display, which is not always a consequence of bsc#1168979, How can you be ok with it when it can be done right?

**Comment by okurz** [comment]
@SergioAtSUSE see https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/10220 for an approach to disable the screensaver

**Comment by foursixnine** [comment]
> @foursixnine, as you are concerned about the time invested (wasted?) in this ticket. Why don't better start working on https://progress.opensuse.org/issues/40490
> 
> When that ticket is done, wouldn't you consider the effort put in this PR a waste of time?

No. As I did already mention: If somebody doesn't schedule the screen saver module, person might get hit by the problem. Therefore, it is not a waste of time.

**Comment by foursixnine** [comment]
> > As for detecting the old outdated screen buffer:
> > ```
> > 1. See the clock: https://openqa.suse.de/tests/4211435#step/boot_to_desktop/9
> >    2: See the clock: https://openqa.suse.de/tests/4211435#step/desktop_stuff/12, `generic-desktop` matches, from the **OLD** (See line above) screen, test presses `esc` and then `alt-f2` to get the desktop runner, since next `check_screen` is for `desktop-runner`, fails and the `else` condition in the code kicks in
> > ```
> 
> I agree. I am telling this same thing since I created the ticket.
> 
> > 3: Since there's nothing to do, the else jumps next, in my code changes, a `record_soft_failure` is triggered  https://openqa.suse.de/tests/4211435#step/desktop_stuff/14 in this case, specifically for aarch64 (but tbh, it is not needed, since it could happen somewhere else).
> 
> the turned off display can happened somewhere else and for any other reason which is not bsc#1168979. So, please don't record a soft-fail for bsc#1168979 when bsc#1168979 is not reproduced.

So let's treat is as a soft reference (L95), and we can then call it a day

File: lib/x11utils.pm (hunk 4 above)
Comment (foursixnine) [reply lib/x11utils.pm:95]: only for aarch64, and again, it's not stating that it is the actual reason, but for aarch64, for the time being, we _know_ that said bug, it's the cause for the behavior.

**Comment by SergioAtSUSE** [comment]
> No. As I did already mention: If somebody doesn't schedule the screen saver module, person might get hit by the problem. Therefore, it is not a waste of time.

This is a point when we don't agree about what is a waste.
You consider a waste something that gives more debugging information.
I consider a waste handling an exception that can be handle with a module.


> I am annoyed at the time I've had to waste explaining why the workaround that you wanted, was a bad idea.

The same for me explaining why the record_soft_fail is wrong in this PR.

I can agree to the implementation shown in #10064 of the workaround being a bad idea. But, the pseudocode for the workaround I suggested is still the only valid workaround for that bug fulfilling the 3 rules:

1. A workaround is avoiding a test to fail because of a product bug
2. A workaround always must record a soft-fail
3. A soft-fail only can be recorded if the bug was reproduced. (not if it could be reproduced)

File: lib/x11utils.pm (hunk 2 above)
Comment (SergioAtSUSE) [inline lib/x11utils.pm:124]: If you remove that `record_soft_failure`, I can approve.
This will record a soft fail anytime the desktop runner is not shown, even when alt-f2 keys a missing.

**Comment by foursixnine** [comment]
> > No. As I did already mention: If somebody doesn't schedule the screen saver module, person might get hit by the problem. Therefore, it is not a waste of time.
> 
> This is a point when we don't agree about what is a waste.
> You consider a waste something that gives more debugging information.
> I consider a waste handling an exception that can be handled with a module.
> 
> > I am annoyed at the time I've had to waste explaining why the workaround that you wanted, was a bad idea.
> 
> The same for me explaining why the record_soft_fail is wrong in this PR.
> 
> I can agree to the implementation shown in #10064 of the workaround being a bad idea. But, the pseudocode for the workaround I suggested is still the only valid workaround for that bug fulfilling the 3 rules:
> 
>     1. A workaround is avoiding a test to fail because of a product bug
> 
>     2. A workaround always must record a soft-fail
> 
>     3. A soft-fail only can be recorded if the bug was reproduced. (not if it could be reproduced)

Make the code sggestions that you _think_ should be, and let's see... but at this point, you can pick up the PR or submit an alternative, I'm tired of trying to convince you.

File: lib/x11utils.pm (hunk 4 above)
Comment (SergioAtSUSE) [reply lib/x11utils.pm:95]: > only for aarch64, and again, it's not stating that it is the actual reason, but for aarch64, for the time being, we _know_ that said bug, it's the cause for the behavior.

Let me use an example of why this approach is wrong.

1. We _know_ that one sympton of corona virus is fiber.
2. Let's diagnose every people with fiber as infected by coronavirus.

**Comment by SergioAtSUSE** [comment]
I think @okurz already agrees with the PR.
And @ggardet didn't add any other concern.

Waiting for travis before merging.

File: lib/x11utils.pm (hunk 4 above)
Comment (foursixnine) [reply lib/x11utils.pm:95]: > > only for aarch64, and again, it's not stating that it is the actual reason, but for aarch64, for the time being, we _know_ that said bug, it's the cause for the behavior.
> 
> Let me use an example of why this approach is wrong.
> 
>     1. We _know_ that one sympton of corona virus is fiber.
> 
>     2. Let's diagnose every people with fiber as infected by coronavirus.

More like 'It might be' corona :D so let's be careful

File: lib/x11utils.pm [hunk 8]
```diff
@@ -112,6 +120,8 @@ sub ensure_unlocked_desktop {
                     assert_screen 'generic-desktop';
                 }
                 else {
+                    diag("Next loop ($counter), Generic desktop didn't match");
+                    record_info('Screen seems frozen', 'Might be consequence of bsc#1168979') if is is_aarch64;
```
Comment (ggardet) [inline lib/x11utils.pm:124]: The additional `is` in `if is is_aarch64` looks like a typo, no?

Comment (foursixnine) [reply lib/x11utils.pm]: Yeah, just fixed it

Comment (SergioAtSUSE) [reply lib/x11utils.pm]: Nice catch!

Comment (SergioAtSUSE) [reply lib/x11utils.pm]: Good that I waited for travis: https://travis-ci.org/github/os-autoinst/os-autoinst-distri-opensuse/jobs/685624978#L16274

**Comment by foursixnine** [comment]
@ggardet just created https://github.com/os-autoinst/os-autoinst-needles-opensuse/pull/661

**Comment by SergioAtSUSE** [comment]
Waiting for needle to be merged
